### PR TITLE
update mortgage logic

### PR DESCRIPTION
**Pull request by nimishagarwal76**

### CODE REVIEW

File: migrations/20170203131246_create_mortgage_details_table.up.sql [hunk 1]
```diff
@@ -3,6 +3,7 @@ CREATE TABLE IF NOT EXISTS MortgageDetails (
 	userId int(11) UNSIGNED NOT NULL,
 	stockId int(11) UNSIGNED NOT NULL,
 	stocksInBank int(11) UNSIGNED NOT NULL,
+	mortgagePrice int(11) UNSIGNED NOT NULL,
```
Comment (PaiAkshay998) [inline migrations/20170203131246_create_mortgage_details_table.up.sql]: Don't modify old migrations. Create a new migration to add the column

File: models/User.go [hunk 2]
```diff
@@ -12,12 +12,12 @@ import (
 	"sync"
 	"sync/atomic"
 
-	"github.com/delta/dalal-street-server/proto_build/datastreams"
+	datastreams_pb "github.com/delta/dalal-street-server/proto_build/datastreams"
```
Comment (PaiAkshay998) [inline models/User.go]: Why was this changed again?

File: models/User.go [hunk 3]
```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
...
-		if user.Total > lim {
-			return nil, WayTooMuchCashError{}
-		}
+		l.Debugf("Mortgaging stocks in action")
+
+		db := getDB()
 
-		rate = MORTGAGE_DEPOSIT_RATE
-		l.Debugf("stockQuantity negative. Depositing stocks to mortgage @ %d%%", rate)
+		/* We don't want to include Mortgage type transaction because mortgage transactions are stored in Transactions
```
Comment (thakkarparth007) [inline models/User.go]: I don't think it's a good idea to make this assumption. Throughout the code we have made one important invariant - `Select sum(StockQuantity) From Transactions where UserId=1 and StockId=2` should **always** give stocks that he currently owns, regardless of the type of the transaction (similar case for `Total` field in the transaction).

While the assumption you're making is consistent with the above "rule", I still feel it's not aligned with it exactly (which is why you're making two queries below to get unmortgaged stocks).

File: models/User.go [hunk 4]
```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
...
+				maxStocksRetrieval = stocksInBank
+			}
+
+			if maxStocksRetrieval > tempStockQuantity { // We don't want to retrieve more than required
+				maxStocksRetrieval = tempStockQuantity
+			}
+			expense := -int64(price) * int64(maxStocksRetrieval) * MORTGAGE_RETRIEVE_RATE / 100
+			trTotal += expense
+			if maxStocksRetrieval < stocksInBank {
+				sql := "UPDATE MortgageDetails SET stocksInBank=? where id=?"
```
Comment (thakkarparth007) [inline models/User.go]: I'm not sure if this will work, it might, I'm just not sure. The for loop is operating on the stream of results of the Select query. Updating the same table during this *might* be unsafe (make sure it's not by reading up or experimenting)

File: models/User.go [hunk 5]
```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
...
+	var trTotal int64
+
+	/* It is retrive action; This block will set correct trTotal and stockQuantity equal
+	to amount of stocks that could be retrieved; This block also modifies MortgageDetails
+	table, btw this table is not used anywhere except in current function */
+
+	if stockQuantity >= 0 {
+
+		db := getDB()
+		sql := "SELECT id, stocksInBank, mortgagePrice from MortgageDetails where userId=? AND stockId=? ORDER BY mortgagePrice"
```
Comment (thakkarparth007) [inline models/User.go]: Don't we want to go in descending order of `mortgagePrice`? Makes more sense, since as a lender, you want to get money for higher valued assets first.

File: models/User.go [hunk 6]
```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
...
+		for rows.Next() {
+			rows.Scan(&id, &stocksInBank, &price)
+			maxStocksRetrieval = tempUserCash / price
+
+			if maxStocksRetrieval == 0 || tempStockQuantity == 0 {
+				stockQuantity -= tempStockQuantity // Amount of stocks retrieved till now
+				break
+			}
+
+			if maxStocksRetrieval > stocksInBank {
```
Comment (thakkarparth007) [inline models/User.go]: I think a helper `max(a, b...)` would be useful here.

File: models/User.go (hunk 4 above)
Comment (thakkarparth007) [reply models/User.go]: Oh, also, there's a bigger problem.

Do not ever make modifications directly to the database using `db.Exec()`. Always form a transaction (database transaction), and commit or rollback it. Otherwise you'll end up in a huge mess (e.g. if server crashes due to bug in some other part of code, and this goroutine is halfway done, like you updated MortgageDetails but not User's cash which happens in another SQL exec)

File: models/User.go [hunk 7]
```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
...
+			if maxStocksRetrieval < stocksInBank {
+				sql := "UPDATE MortgageDetails SET stocksInBank=? where id=?"
+				err = db.Exec(sql, stocksInBank-maxStocksRetrieval, id).Error
+				if err != nil {
+					l.Infof("PerformMortgageTransaction failed for userId = %d, stockId = %d amount = %d while updating retrieved stocks", userId, stockId, stockQuantity)
+					l.Error(err)
+					return nil, err
+				}
+			} else /* So we can delete that entire row */ {
+				sql := "DELETE from MortgageDetails WHERE id=?"
```
Comment (thakkarparth007) [inline models/User.go]: If you delete from here, and you don't update the Transactions table, the invariant I mentioned earlier gets violated. Secondly, we don't have any trail to follow later if we need to understand what happened.

File: models/User.go (hunk 5 above)
Comment (Harsh2098) [reply models/User.go]: player would want to retrieve maximum stocks out of mortgage with less cash, so retirve stocks with less mortgage price first

File: models/User.go (hunk 7 above)
Comment (Harsh2098) [reply models/User.go]: Using database transaction now

File: migrations/20180215163030_add_mortgage_price.up.sql [hunk 8]
```diff
@@ -0,0 +1 @@
+ALTER TABLE MortgageDetails ADD COLUMN mortgagePrice int(11);
```
Comment (PaiAkshay998) [inline migrations/20180215163030_add_mortgage_price.up.sql]: Where is the corresponding .down file?

Comment (PaiAkshay998) [reply migrations/20180215163030_add_mortgage_price.up.sql]: And since you're editing might as well add a new line at the end of the up file :stuck_out_tongue:

File: models/User.go [hunk 9]
```diff
@@ -1394,11 +1394,12 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
 		"param_stockQuantity": stockQuantity,
 	})
 
-	l.Infof("PerformMortgageTransaction requested")
+	l.Infof("PerformMortgageTransaction requested for userId = %d, stockId = %d amount = %d", userId, stockId, stockQuantity)
 
 	l.Debugf("Acquiring exclusive write on user")
 	ch, user, err := getUserExclusively(userId)
 	if err != nil {
+		l.Infof("PerformMortgageTransaction failed for userId = %d, stockId = %d amount = %d while retrieving user", userId, stockId, stockQuantity)
```
Comment (PaiAkshay998) [inline models/User.go]: This seems redundant. Could just alter the error log in the next line.

File: migrations/20180215163030_add_mortgage_price.up.sql (hunk 8 above)
Comment (Harsh2098) [reply migrations/20180215163030_add_mortgage_price.up.sql]: Done

File: models/User.go [hunk 10]
```diff
@@ -1408,75 +1408,147 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
...
+
+		// tempStockQuantity is stock quantity left to retrieve
+		tempStockQuantity := stockQuantity
+		tempUserCash := int64(user.Cash)
+		l.Infof("Cash %d", tempUserCash)
+		var maxStocksRetrieval int64
+
+		for rows.Next() {
+			rows.Scan(&id, &stocksInBank, &price)
+			maxStocksRetrieval = tempUserCash / price
```
Comment (PaiAkshay998) [inline models/User.go]: I feel this logic can be simplified a lot. Breaking out of loop etc feels like a complication that can be avoided. A simple way to have done this would have been

```python
while stocks_to_retrieve > 0:
    id, stocks_in_bank, price = retrieve_record_from_db()
    if stocks_in_bank < stocks_to_retrieve:
        ...
    elif stocks_in_bank > stocks_to_retrieve:
        ...
    else stocks_in_bank == stocks_to_retrieve:
        ...
```
I mean sure the method you've used might work too. But incase something goes wrong later, I feel this is easier to pinpoint.

File: models/MortgageDetail.go [hunk 11]
```diff
@@ -23,8 +27,144 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		return 0, err
+	}
+
+	var stocksInBank int64
+
+	if !rows.Next() {
+		sql := "INSERT into MortgageDetails (userId, stockId, stocksInBank, mortgagePrice) VALUES (?, ?, ?, ?)"
+
+		// Here mortgage price is last average price
+		err = tx.Exec(sql, user.Id, stockID, -stockQuantity, mortgagePrice).Error
```
[the remark this answers is not in the document]
Comment (Harsh2098) [reply models/MortgageDetail.go]: done

File: models/MortgageDetail.go [hunk 12]
```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		return 0, err
+	}
+
+	if stockQuantity > stockOwned {
+		l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockOwned, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	stocksInBank, err := getStocksInBank(user.Id, stockID, mortgagePrice, tx)
+	if (err != nil && err != InvalidRetrievePriceError{} /* This type of error not applicable here */) {
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: Comment outside brackets

File: models/MortgageDetail.go [hunk 13]
```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		sql := "INSERT into MortgageDetails (userId, stockId, stocksInBank, mortgagePrice) VALUES (?, ?, ?, ?)"
+		err = tx.Exec(sql, user.Id, stockID, stockQuantity, mortgagePrice).Error
+	} else {
+		sql := "UPDATE MortgageDetails SET stocksInBank=? WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, stocksInBank+stockQuantity, user.Id, stockID, mortgagePrice).Error
+	}
+
+	if err != nil {
+		l.Error(err)
+		tx.Rollback()
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: Let the caller rollback, since the caller is the one that created `tx`

File: models/MortgageDetail.go [hunk 14]
```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		"param_stockId":       stockID,
+		"param_stockQuantity": stockQuantity,
+		"param_retrievePrice": retrievePrice,
+	})
+
+	l.Debugf("Retrieving stocks in action")
+
+	stocksInBank, err := getStocksInBank(userID, stockID, retrievePrice, tx)
+	if err != nil {
+		l.Infof("retrieveStocksActions failed for userId = %d, stockId = %d amount = %d while retrieving mortgaged stocks", userID, stockID, stockQuantity)
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: Remove this log. `l.Error(...)` will print the same information. userID, stockID, stockQty etc will be printed because we've created `l` `WithFields` to print those for every statement it prints.

File: models/MortgageDetail.go [hunk 15]
```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		l.Error(err)
+		tx.Rollback()
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: Remove `tx.Rollback()`. Let caller roll back the tx.

File: models/MortgageDetail.go [hunk 16]
```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+
+	} else /* stockQuantity to retrieve > stocksInBank */ {
+		l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stocksInBank, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	if err != nil {
+		l.Infof("retrieveStocksAction failed for userId = %d, stockId = %d amount = %d deleting retrieved user", userID, stockID, stockQuantity)
+		l.Error(err)
+		tx.Rollback()
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: Let caller rollback the tx.

File: models/MortgageDetail.go [hunk 17]
```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		sql := "DELETE from MortgageDetails WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, userID, stockID, retrievePrice).Error
+
+	} else /* stockQuantity to retrieve > stocksInBank */ {
+		l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stocksInBank, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	if err != nil {
+		l.Infof("retrieveStocksAction failed for userId = %d, stockId = %d amount = %d deleting retrieved user", userID, stockID, stockQuantity)
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: Remove this. l.Error(...) conveys same info.

File: models/User.go [hunk 18]
```diff
@@ -1386,20 +1393,22 @@ func (e WayTooMuchCashError) Error() string {
...
 	var l = logger.WithFields(logrus.Fields{
 		"method":              "PerformMortgageTransaction",
 		"param_userId":        userId,
 		"param_stockId":       stockId,
 		"param_stockQuantity": stockQuantity,
+		"param_retrievePrice": retrievePrice,
 	})
 
-	l.Infof("PerformMortgageTransaction requested")
+	l.Infof("PerformMortgageTransaction requested for userId = %d, stockId = %d amount = %d retrievePrice = %d", userId, stockId, stockQuantity, retrievePrice)
```
Comment (thakkarparth007) [inline models/User.go]: Remove the parameters, they'll be printed automatically because `l` is created `WithFields` to print those fields for every statement l prints.

File: models/MortgageDetail.go [hunk 19]
```diff
@@ -23,8 +26,122 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		return 0, err
+	}
+
+	if stockQuantity > stockOwned {
+		l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockOwned, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	stocksInBank, err := getStocksInBank(user.Id, stockID, mortgagePrice, tx)
+	if (err != nil && err != InvalidRetrievePriceError{}) /* This type of error not applicable here */ {
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: comment after {

File: models/MortgageDetail.go [hunk 20]
```diff
@@ -23,8 +26,122 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+	if int64(userCash/retrievePrice) < stockQuantity {
+		l.Errorf("Insufficient cash with user. Have %d, want %d", userCash, stockQuantity*int64(retrievePrice))
+		return 0, NotEnoughCashError{}
+	}
+
+	if stockQuantity < stocksInBank {
+		sql := "UPDATE MortgageDetails SET stocksInBank=? WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, stocksInBank-stockQuantity, userID, stockID, retrievePrice).Error
+
+	} else if stockQuantity == stocksInBank /* So we can delete that entire row */ {
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: comment after {

File: models/MortgageDetail.go [hunk 21]
```diff
@@ -23,8 +26,122 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		sql := "DELETE from MortgageDetails WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, userID, stockID, retrievePrice).Error
+
+	} else /* stockQuantity to retrieve > stocksInBank */ {
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: comment after {

File: models/MortgageDetail.go [hunk 22]
```diff
@@ -23,8 +26,122 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	return -int64(retrievePrice) * stockQuantity * MORTGAGE_RETRIEVE_RATE / 100, nil
+}
+
+func getStocksInBank(userID, stockID uint32, retrievePrice uint64, tx *gorm.DB) (int64, error) {
+
```
Comment (thakkarparth007) [inline models/MortgageDetail.go]: remove empty line, it's start of fn

File: models/MortgageDetail.go (hunk 19 above)
Comment (Harsh2098) [reply models/MortgageDetail.go]: D

File: models/MortgageDetail.go (hunk 20 above)
Comment (Harsh2098) [reply models/MortgageDetail.go]: O

File: models/MortgageDetail.go (hunk 21 above)
Comment (Harsh2098) [reply models/MortgageDetail.go]: N

File: models/MortgageDetail.go (hunk 22 above)
Comment (Harsh2098) [reply models/MortgageDetail.go]: E

File: models/MortgageDetail.go [hunk 23]
```diff
@@ -23,8 +26,121 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+	l.Debugf("Retrieving stocks in action")
+
+	stocksInBank, err := getStocksInBank(userID, stockID, retrievePrice, tx)
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	if int64(userCash/retrievePrice) < stockQuantity {
```
Comment (PaiAkshay998) [inline models/MortgageDetail.go]: Why not ```userCash` < stockQuantity * retrievePrice```?

File: models/MortgageDetail.go (hunk 19 above)
Comment (thakkarparth007) [reply models/MortgageDetail.go]: Y

File: models/MortgageDetail.go (hunk 20 above)
Comment (thakkarparth007) [reply models/MortgageDetail.go]: A

File: models/MortgageDetail.go (hunk 21 above)
Comment (thakkarparth007) [reply models/MortgageDetail.go]: Y

File: models/MortgageDetail.go (hunk 22 above)
Comment (thakkarparth007) [reply models/MortgageDetail.go]: Y

File: models/MortgageDetail.go [hunk 24]
```diff
@@ -23,8 +26,121 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
...
+		l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stocksInBank, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	return -int64(retrievePrice) * stockQuantity * MORTGAGE_RETRIEVE_RATE / 100, nil
```
Comment (PaiAkshay998) [inline models/MortgageDetail.go]: So say a stock is worth Rs. 100.
If I mortgage a stock, I get Rs. 80 (80%) back and when I buy the stock back I pay Rs. 90 (90%)?

Comment (Harsh2098) [reply models/MortgageDetail.go]: Yes

Comment (Harsh2098) [reply models/MortgageDetail.go]: You are right

File: models/MortgageDetail.go (hunk 23 above)
Comment (PaiAkshay998) [reply models/MortgageDetail.go]: @Harsh2098 this also please.

**Comment by PaiAkshay998** [comment]
Awesome :fire:
